fili-core/src/main/java/com/yahoo/bard/webservice/web/responseprocessors/DruidPartialDataResponseProcessor.java

```diff
@@ -0,0 +1,261 @@
+     *         DruidPartialDataResponseProcessor::processResponse,
+     *     </li>
+     *     <li>
+     *         Parse both the uncoveredIntervals extracted above and allAvailableIntervals extracted from the union of
+     *         all the query's datasource's availabilities from DataSourceMetadataService into SimplifiedIntervalLists,
+     *     </li>
+     *     <li>
+     *         Compare both SimplifiedIntervalLists above, if allAvailableIntervals has any overlap with
+     *         uncoveredIntervals, invoke error response indicating druid is missing some data that are we are expecting
+     *         to exists.
+     *     </li>
+     * </ol>
+     *
+     * @param json  The json representing a druid data response
+     * @param query  The query with the schema for processing this response
+     * @param metadata  The LoggingContext to use
+     */
+    @Override
+    public void processResponse(JsonNode json, DruidAggregationQuery<?> query, LoggingContext metadata) {
+        validateJsonResponse(json, query);
+
+        int statusCode = json.get(DruidJsonResponseContentKeys.STATUS_CODE.getName()).asInt();
+        if (statusCode == Status.OK.getStatusCode()) {
+            checkOverflow(json, query);
+
+            SimplifiedIntervalList overlap = getOverlap(json, query);
+            if (!overlap.isEmpty()) {
+                logAndGetErrorCallback(ErrorMessageFormat.DATA_AVAILABILITY_MISMATCH.format(overlap), query);
+            }
+            if (next instanceof FullResponseProcessor) {
+                next.processResponse(json, query, metadata);
+            } else {
+                next.processResponse(json.get(DruidJsonResponseContentKeys.RESPONSE.getName()), query, metadata);
+            }
+        } else if (statusCode == Status.NOT_MODIFIED.getStatusCode() && !(next instanceof FullResponseProcessor)) {
+            logAndGetErrorCallback(
+                    "Content Not Modified(304), but no etag cache response processor is available to process " +
+                            "the 304 response",
+                    query);
+        } else {
+            next.processResponse(json, query, metadata);
+        }
+    }
+
+    /**
+     * Validates JSON response object to make sure it contains all of the following information.
+     * <ul>
+     *     <li>X-Druid-Response-Context
+     *         <ol>
+     *             <li>uncoveredIntervals</li>
+     *             <li>uncoveredIntervalsOverflowed</li>
+     *         </ol>
+     *     </li>
+     *     <li>status-code</li>
+     * </ul>
+     *
+     * @param json  The JSON response that is to be validated
+     * @param query  The query with the schema for processing this response
+     */
+    private void validateJsonResponse(JsonNode json, DruidAggregationQuery<?> query) {
+        if (json.getNodeType() == JsonNodeType.ARRAY) {
+            logAndGetErrorCallback("JSON response is missing X-Druid-Response-Context and status code", query);
+        }
```

Comment:
All of these error strings should be moved into constants and referenced by the constants in the tests so that they'll stay in sync.